### ISSUE TITLE
The multiceiver demo not working well

### ISSUE THREAD

**Issue by leleleoo123** [TMRh20/nrf_to_nrf#2]
Dear @TMRh20, thanks for sharing this wonderful library!
I try to port the multiceiver demo in RF24 repo, change the code following your instructions in readme. I set one seeduino xiao ble to be receiver, and other two xiao bles as transmitters, but don't work well. The receiver successfully received most messages from node#0, but lost most messages from node#1, and node#1 prints `transmission failed` very very often.

**Comment by leleleoo123** [comment]
I tried to use esp32+nrf24l01 to replace the xiao ble as receiver, still same issue.

**Comment by 2bndy5** [comment]
> change the code following your instructions in readme

For clarity, the important part in the readme is using dynamic payloads (on all nodes).

**Comment by leleleoo123** [comment]
Yes I did use the dynamic payloads @2bndy5 。Only after I turn off the node#0, the packets from node#1 start to be received ok.

**Comment by 2bndy5** [comment]
Did you make other changes to the example without mentioning?

I say this because you have to let the nodes delay between transmissions. Otherwise, only 1 node will hog the channel and basically disallow other transmissions to get through.

**Comment by leleleoo123** [comment]
@2bndy5 No, I only made changes as mentioned in readme, I didn't change the 1000ms transmission delay.

**Comment by 2bndy5** [comment]
That's very strange. The multicieverDemo is basically a super simple version of what RF24Network lib is doing with addresses. TMRh20 reported RF24Network working well, so I'm rather stumped now. Thank you for tolerating my questions.

**Comment by TMRh20** [comment]
I just tested the example between my XIAO BLE and 2 RPis and it worked fine. I am also kind of stumped because it tests OK on my end. 

Here is how the code should look on the XIAO. Note the changes to
`uint8_t bytes = radio.getDynamicPayloadSize(); `
and   `radio.enableDynamicPayloads();`


```
/*
 * See documentation at https://nRF24.github.io/RF24
 * See License information at root directory of this library
 * Author: Brendan Doherty 2bndy5
 */

/**
 * A simple example of sending data from as many as 6 nRF24L01 transceivers to
 * 1 receiving transceiver. This technique is trademarked by
 * Nordic Semiconductors as "MultiCeiver".
 *
 * This example was written to be used on up to 6 devices acting as TX nodes &
 * only 1 device acting as the RX node (that's a maximum of 7 devices).
 * Use the Serial Monitor to change each node's behavior.
 */
#include <SPI.h>
#include "printf.h"
//#include "RF24.h"
#include "nrf_to_nrf.h"
// instantiate an object for the nRF24L01 transceiver
//RF24 radio(7, 8);  // using pin 7 for the CE pin, and pin 8 for the CSN pin
nrf_to_nrf radio;

// For this example, we'll be using 6 addresses; 1 for each TX node
// It is very helpful to think of an address as a path instead of as
// an identifying device destination
// Notice that the last byte is the only byte that changes in the last 5
// addresses. This is a limitation of the nRF24L01 transceiver for pipes 2-5
// because they use the same first 4 bytes from pipe 1.
uint64_t address[6] = { 0x7878787878LL,
                        0xB3B4B5B6F1LL,
                        0xB3B4B5B6CDLL,
                        0xB3B4B5B6A3LL,
                        0xB3B4B5B60FLL,
                        0xB3B4B5B605LL };

// Because this example allow up to 6 nodes (specified by numbers 0-5) to
// transmit and only 1 node to receive, we will use a negative value in our
// role variable to signify this node is a receiver.
// role variable is used to control whether this node is sending or receiving
char role = 'R';  // integers 0-5 = TX node; character 'R' or integer 82 = RX node

// For this example, we'll be using a payload containing
// a node ID number and a single integer number that will be incremented
// on every successful transmission.
// Make a data structure to use as a payload.
struct PayloadStruct {
  unsigned long nodeID;
  unsigned long payloadID;
};
PayloadStruct payload;

// This example uses all 6 pipes to receive while TX nodes only use 2 pipes
// To make this easier we'll use a function to manage the addresses, and the
// payload's nodeID
void setRole();  // declare a prototype; definition is found after the loop()

void setup() {

  Serial.begin(115200);
  while (!Serial) {
    // some boards need to wait to ensure access to serial over USB
  }

  // initialize the transceiver on the SPI bus
  if (!radio.begin()) {
    Serial.println(F("radio hardware is not responding!!"));
    while (1) {}  // hold in infinite loop
  }

  // print example's introductory prompt
  Serial.println(F("RF24/examples/MulticeiverDemo"));
  Serial.println(F("*** Enter a number between 0 and 5 (inclusive) to change"));
  Serial.println(F("    the identifying node number that transmits."));

  // Set the PA Level low to try preventing power supply related problems
  // because these examples are likely run with nodes in close proximity of
  // each other.
  radio.setPALevel(NRF_PA_MAX);  // RF24_PA_MAX is default.

  // save on transmission time by setting the radio to only transmit the
  // number of bytes we need to transmit a float
  //radio.setPayloadSize(sizeof(payload));  // 2x int datatype occupy 8 bytes
  radio.enableDynamicPayloads();

  // Set the pipe addresses accordingly. This function additionally also
  // calls startListening() or stopListening() and sets the payload's nodeID
  setRole();

  // For debugging info
  // printf_begin();             // needed only once for printing details
  // radio.printDetails();       // (smaller) function that prints raw register values
  // radio.printPrettyDetails(); // (larger) function that prints human readable data

}  // setup()

void loop() {

  if (role <= 53) {
    // This device is a TX node

    unsigned long start_timer = micros();                  // start the timer
    bool report = radio.write(&payload, sizeof(payload));  // transmit & save the report
    unsigned long end_timer = micros();                    // end the timer

    if (report) {
      // payload was delivered

      Serial.print(F("Transmission of payloadID "));
      Serial.print(payload.payloadID);  // print payloadID
      Serial.print(F(" as node "));
      Serial.print(payload.nodeID);  // print nodeID
      Serial.print(F(" successful!"));
      Serial.print(F(" Time to transmit: "));
      Serial.print(end_timer - start_timer);  // print the timer result
      Serial.println(F(" us"));
    } else {
      Serial.println(F("Transmission failed or timed out"));  // payload was not delivered
    }
    payload.payloadID++;  // increment payload number

    // to make this example readable in the serial monitor
    delay(1000);  // slow transmissions down by 1 second

  } else if (role == 'R') {
    // This device is the RX node

    uint8_t pipe;
    if (radio.available(&pipe)) {              // is there a payload? get the pipe number that recieved it
      uint8_t bytes = radio.getDynamicPayloadSize();  // get the size of the payload
      radio.read(&payload, bytes);             // fetch payload from FIFO
      Serial.print(F("Received "));
      Serial.print(bytes);  // print the size of the payload
      Serial.print(F(" bytes on pipe "));
      Serial.print(pipe);  // print the pipe number
      Serial.print(F(" from node "));
      Serial.print(payload.nodeID);  // print the payload's origin
      Serial.print(F(". PayloadID: "));
      Serial.println(payload.payloadID);  // print the payload's number
    }
  }  // role

  if (Serial.available()) {
    // change the role via the serial monitor

    char c = Serial.read();
    if (toupper(c) == 'R' && role <= 53) {
      // Become the RX node

      role = 'R';
      Serial.println(F("*** CHANGING ROLE TO RECEIVER ***"));
      Serial.println(F("--- Enter a number between 0 and 5 (inclusive) to act as"));
      Serial.println(F("    a unique node number that transmits to the RX node."));
      setRole();  // change address on all pipes to TX nodes

    } else if (c >= 48 && c <= 53 && c != role) {
      // Become a TX node with identifier 'c'

      role = c - 48;
      Serial.print(F("*** CHANGING ROLE TO NODE "));
      Serial.print(c);
      Serial.println(F(" ***"));
      Serial.println(F("--- Enter a number between 0 and 5 (inclusive) to change"));
      Serial.println(F("    the identifying node number that transmits."));
      Serial.println(F("--- PRESS 'R' to act as the RX node."));
      setRole();  // change address on pipe 0 to the RX node
    }
  }

}  // loop

void setRole() {
  if (role == 'R') {
    // For the RX node

    // Set the addresses for all pipes to TX nodes
    for (uint8_t i = 0; i < 6; ++i)
      radio.openReadingPipe(i, address[i]);

    radio.startListening();  // put radio in RX mode

  } else {
    // For the TX node

    // set the payload's nodeID & reset the payload's identifying number
    payload.nodeID = role;
    payload.payloadID = 0;

    // Set the address on pipe 0 to the RX node.
    radio.stopListening();  // put radio in TX mode
    radio.openWritingPipe(address[role]);

    // According to the datasheet, the auto-retry features's delay value should
    // be "skewed" to allow the RX node to receive 1 transmission at a time.
    // So, use varying delay between retry attempts and 15 (at most) retry attempts
    radio.setRetries(((role * 3) % 12) + 3, 15);  // maximum value is 15 for both args
  }
}  // setRole
```

**Comment by leleleoo123** [comment]
@TMRh20 Yes, very strange, my code was the same with yours except that I set the `role` in code, instead of reading from serial to change role:
```c++
/*
 * See documentation at https://nRF24.github.io/RF24
 * See License information at root directory of this library
 * Author: Brendan Doherty 2bndy5
 */

/**
 * A simple example of sending data from as many as 6 nRF24L01 transceivers to
 * 1 receiving transceiver. This technique is trademarked by
 * Nordic Semiconductors as "MultiCeiver".
 *
 * This example was written to be used on up to 6 devices acting as TX nodes &
 * only 1 device acting as the RX node (that's a maximum of 7 devices).
 * Use the Serial Monitor to change each node's behavior.
 */
#include <SPI.h>
#include "printf.h"
//#include "RF24.h"
#include "nrf_to_nrf.h"
// instantiate an object for the nRF24L01 transceiver
//RF24 radio(7, 8);  // using pin 7 for the CE pin, and pin 8 for the CSN pin
nrf_to_nrf radio;

// For this example, we'll be using 6 addresses; 1 for each TX node
// It is very helpful to think of an address as a path instead of as
// an identifying device destination
// Notice that the last byte is the only byte that changes in the last 5
// addresses. This is a limitation of the nRF24L01 transceiver for pipes 2-5
// because they use the same first 4 bytes from pipe 1.
uint64_t address[6] = { 0x7878787878LL,
                        0xB3B4B5B6F1LL,
                        0xB3B4B5B6CDLL,
                        0xB3B4B5B6A3LL,
                        0xB3B4B5B60FLL,
                        0xB3B4B5B605LL };

// Because this example allow up to 6 nodes (specified by numbers 0-5) to
// transmit and only 1 node to receive, we will use a negative value in our
// role variable to signify this node is a receiver.
// role variable is used to control whether this node is sending or receiving
char role = 'R';  // for receiver
//char role = 0;    // for node#0
//char role = 1;   // for node#1

// For this example, we'll be using a payload containing
// a node ID number and a single integer number that will be incremented
// on every successful transmission.
// Make a data structure to use as a payload.
struct PayloadStruct {
  unsigned long nodeID;
  unsigned long payloadID;
};
PayloadStruct payload;

// This example uses all 6 pipes to receive while TX nodes only use 2 pipes
// To make this easier we'll use a function to manage the addresses, and the
// payload's nodeID
void setRole();  // declare a prototype; definition is found after the loop()

void setup() {

  Serial.begin(115200);
  while (!Serial) {
    // some boards need to wait to ensure access to serial over USB
  }

  // initialize the transceiver on the SPI bus
  if (!radio.begin()) {
    Serial.println(F("radio hardware is not responding!!"));
    while (1) {}  // hold in infinite loop
  }

  // print example's introductory prompt
  Serial.println(F("RF24/examples/MulticeiverDemo"));
  Serial.println(F("*** Enter a number between 0 and 5 (inclusive) to change"));
  Serial.println(F("    the identifying node number that transmits."));

  // Set the PA Level low to try preventing power supply related problems
  // because these examples are likely run with nodes in close proximity of
  // each other.
  radio.setPALevel(NRF_PA_MAX);  // RF24_PA_MAX is default.

  // save on transmission time by setting the radio to only transmit the
  // number of bytes we need to transmit a float
  //radio.setPayloadSize(sizeof(payload));  // 2x int datatype occupy 8 bytes
  radio.enableDynamicPayloads();

  // Set the pipe addresses accordingly. This function additionally also
  // calls startListening() or stopListening() and sets the payload's nodeID
  setRole();

  // For debugging info
  // printf_begin();             // needed only once for printing details
  // radio.printDetails();       // (smaller) function that prints raw register values
  // radio.printPrettyDetails(); // (larger) function that prints human readable data

}  // setup()

void loop() {

  if (role <= 53) {
    // This device is a TX node

    unsigned long start_timer = micros();                  // start the timer
    bool report = radio.write(&payload, sizeof(payload));  // transmit & save the report
    unsigned long end_timer = micros();                    // end the timer

    if (report) {
      // payload was delivered

      Serial.print(F("Transmission of payloadID "));
      Serial.print(payload.payloadID);  // print payloadID
      Serial.print(F(" as node "));
      Serial.print(payload.nodeID);  // print nodeID
      Serial.print(F(" successful!"));
      Serial.print(F(" Time to transmit: "));
      Serial.print(end_timer - start_timer);  // print the timer result
      Serial.println(F(" us"));
    } else {
      Serial.println(F("Transmission failed or timed out"));  // payload was not delivered
    }
    payload.payloadID++;  // increment payload number

    // to make this example readable in the serial monitor
    delay(1000);  // slow transmissions down by 1 second

  } else if (role == 'R') {
    // This device is the RX node

    uint8_t pipe;
    if (radio.available(&pipe)) {              // is there a payload? get the pipe number that recieved it
      uint8_t bytes = radio.getDynamicPayloadSize();  // get the size of the payload
      radio.read(&payload, bytes);             // fetch payload from FIFO
      Serial.print(F("Received "));
      Serial.print(bytes);  // print the size of the payload
      Serial.print(F(" bytes on pipe "));
      Serial.print(pipe);  // print the pipe number
      Serial.print(F(" from node "));
      Serial.print(payload.nodeID);  // print the payload's origin
      Serial.print(F(". PayloadID: "));
      Serial.println(payload.payloadID);  // print the payload's number
    }
  }  // role

}  // loop

void setRole() {
  if (role == 'R') {
    // For the RX node

    // Set the addresses for all pipes to TX nodes
    for (uint8_t i = 0; i < 6; ++i)
      radio.openReadingPipe(i, address[i]);

    radio.startListening();  // put radio in RX mode

  } else {
    // For the TX node

    // set the payload's nodeID & reset the payload's identifying number
    payload.nodeID = role;
    payload.payloadID = 0;

    // Set the address on pipe 0 to the RX node.
    radio.stopListening();  // put radio in TX mode
    radio.openWritingPipe(address[role]);

    // According to the datasheet, the auto-retry features's delay value should
    // be "skewed" to allow the RX node to receive 1 transmission at a time.
    // So, use varying delay between retry attempts and 15 (at most) retry attempts
    radio.setRetries(((role * 3) % 12) + 3, 15);  // maximum value is 15 for both args
  }
}  // setRole
```

**Comment by 2bndy5** [comment]
This is the only example in the RF24 lib that still uses `uint64_t` addresses. Although that suspicion doesn't pan out given the different behaviors mentioned here.

> Only after I turn off the node#0, the packets from node#1 start to be received ok.

This seriously sounds like OTA collision.
1. Are you starting the TX nodes at exactly the same time? This question seems like a long shot as it would likely entail using a USB hub to power the nodes.
2. If you re-enable node#0 after confirming that node#1 transmits, does node#1 stop getting through again?

**Comment by leleleoo123** [comment]
> Are you starting the TX nodes at exactly the same time? This question seems like a long shot as it would likely need a USB hub to power-up the nodes simultaneously.
Not likely, I was using different usb port on my desktop to power-up the xiaoBles. But maybe I can change the delay time of node#0 to 1500 ms to test?

> If you re-enable node#0 after confirming that node#1 transmits, does node#1 stop getting through again?
Not sure, sometimes after I re-enable node#0, node#1 stop getting through again. But sometimes the node#1 still getting through but node#0 can't

**Comment by leleleoo123** [comment]
I try to set the delay() time of node#0 to be 1000ms, and node#1 to be 3700ms. If the node#0 is closer to the receiver, its messages start to get through, and node#1 always fail transmission. Then if I move node#1 closer to the receiver, node#0 starts to always show transmission failed, and messages of node#1 start getting through!

**Comment by leleleoo123** [comment]
So weird, several weeks ago I have used several esp32+nrf24l01 to run the RF24 multiceiver example, it worked very well even when I only delay(10) ms. Maybe something is wrong with xiaoBle.

**Comment by 2bndy5** [comment]
I have 2 XIAO BLE boards. I'll try to reproduce this later today...

**Comment by leleleoo123** [comment]
thanks so much, looking forward to your good news.

**Comment by 2bndy5** [comment]
<details><summary>I'm getting compiler errors using the code from TMRh20</summary>

```
Using board 'xiaonRF52840' from platform in folder: C:\Users\%USERNAME%\AppData\Local\Arduino15\packages\Seeeduino\hardware\nrf52\1.1.1

c:\Users\%USERNAME%\Documents\Arduino\libraries\nrf_to_nrf\src\nrf_to_nrf.cpp: In member function 'bool nrf_to_nrf::startWrite(void*, uint8_t, bool)':
c:\Users\%USERNAME%\Documents\Arduino\libraries\nrf_to_nrf\src\nrf_to_nrf.cpp:278:1: error: no return statement in function returning non-void [-Werror=return-type]
  278 | }
      | ^
c:\Users\%USERNAME%\Documents\Arduino\libraries\nrf_to_nrf\src\nrf_to_nrf.cpp: In member function 'bool nrf_to_nrf::writeAckPayload(uint8_t, const void*, uint8_t)':
c:\Users\%USERNAME%\Documents\Arduino\libraries\nrf_to_nrf\src\nrf_to_nrf.cpp:285:1: error: no return statement in function returning non-void [-Werror=return-type]
  285 | }
      | ^
c:\Users\%USERNAME%\Documents\Arduino\libraries\nrf_to_nrf\src\nrf_to_nrf.cpp: In member function 'void nrf_to_nrf::enableDynamicPayloads(uint8_t)':
c:\Users\%USERNAME%\Documents\Arduino\libraries\nrf_to_nrf\src\nrf_to_nrf.cpp:401:35: warning: suggest parentheses around '-' inside '<<' [-Wparentheses]
  401 |                      (addressWidth-1 << RADIO_PCNF1_BALEN_Pos) |
      |                       ~~~~~~~~~~~~^~
c:\Users\%USERNAME%\Documents\Arduino\libraries\nrf_to_nrf\src\nrf_to_nrf.cpp: In member function 'void nrf_to_nrf::disableDynamicPayloads()':
c:\Users\%USERNAME%\Documents\Arduino\libraries\nrf_to_nrf\src\nrf_to_nrf.cpp:416:35: warning: suggest parentheses around '-' inside '<<' [-Wparentheses]
  416 |                      (addressWidth-1 << RADIO_PCNF1_BALEN_Pos) |
      |                       ~~~~~~~~~~~~^~
c:\Users\%USERNAME%\Documents\Arduino\libraries\nrf_to_nrf\src\nrf_to_nrf.cpp: In member function 'void nrf_to_nrf::setPayloadSize(uint8_t)':
c:\Users\%USERNAME%\Documents\Arduino\libraries\nrf_to_nrf\src\nrf_to_nrf.cpp:432:35: warning: suggest parentheses around '-' inside '<<' [-Wparentheses]
  432 |                      (addressWidth-1 << RADIO_PCNF1_BALEN_Pos) |
      |                       ~~~~~~~~~~~~^~
c:\Users\%USERNAME%\Documents\Arduino\libraries\nrf_to_nrf\src\nrf_to_nrf.cpp: In member function 'void nrf_to_nrf::setAddressWidth(uint8_t)':
c:\Users\%USERNAME%\Documents\Arduino\libraries\nrf_to_nrf\src\nrf_to_nrf.cpp:719:31: warning: suggest parentheses around '-' inside '<<' [-Wparentheses]
  719 |                      (a_width - 1                  << RADIO_PCNF1_BALEN_Pos) |
      |                       ~~~~~~~~^~~
c:\Users\%USERNAME%\Documents\Arduino\libraries\nrf_to_nrf\src\nrf_to_nrf.cpp: In member function 'void nrf_to_nrf::openReadingPipe(uint8_t, const uint8_t*)':
c:\Users\%USERNAME%\Documents\Arduino\libraries\nrf_to_nrf\src\nrf_to_nrf.cpp:10:15: warning: '*((void*)& prefixArray +3)' is used uninitialized in this function [-Wuninitialized]
   10 |       (p_inp[3] << 24) | (p_inp[2] << 16) | (p_inp[1] << 8) | (p_inp[0]);
      |        ~~~~~~~^
c:\Users\%USERNAME%\Documents\Arduino\libraries\nrf_to_nrf\src\nrf_to_nrf.cpp:10:34: warning: '*((void*)& prefixArray +2)' is used uninitialized in this function [-Wuninitialized]
   10 |       (p_inp[3] << 24) | (p_inp[2] << 16) | (p_inp[1] << 8) | (p_inp[0]);
      |                           ~~~~~~~^
c:\Users\%USERNAME%\Documents\Arduino\libraries\nrf_to_nrf\src\nrf_to_nrf.cpp:10:53: warning: '*((void*)& prefixArray +1)' is used uninitialized in this function [-Wuninitialized]
   10 |       (p_inp[3] << 24) | (p_inp[2] << 16) | (p_inp[1] << 8) | (p_inp[0]);
      |                                              ~~~~~~~^
cc1plus.exe: some warnings being treated as errors

Using library SPI at version 1.0 in folder: C:\Users\%USERNAME%\AppData\Local\Arduino15\packages\Seeeduino\hardware\nrf52\1.1.1\libraries\SPI 
Using library RF24 at version 1.4.6 in folder: C:\Users\%USERNAME%\Documents\Arduino\libraries\RF24 
Using library nrf_to_nrf at version 1.0 in folder: C:\Users\%USERNAME%\Documents\Arduino\libraries\nrf_to_nrf 
Using library Adafruit TinyUSB Library at version 1.7.0 in folder: C:\Users\%USERNAME%\AppData\Local\Arduino15\packages\Seeeduino\hardware\nrf52\1.1.1\libraries\Adafruit_TinyUSB_Arduino
```

</details>

Its worth noting that I always set the Arduino IDE (currently using v2.0.3) to show me compiler warnings. The errors that crash the compiler are about 
https://github.com/TMRh20/nrf_to_nrf/blob/53fc035564bb994dbdc5dd5c57cb0e98a7b517b4/src/nrf_to_nrf.cpp#L274-L285
Since there's no TX/RX FIFOs implemented in this lib, the functions should arbitrarily return true (indicating that the payload is ready to transmit). Using this fix, I can get the example to compile (despite the numerous warnings).

---
It looks like the seeduino nRF52 core has seen some updates recently, but it I don't see anything significant since this lib was written. https://github.com/Seeed-Studio/Adafruit_nRF52_Arduino/compare/v1.0.0...v1.1.1

Hardware testing is still forthcoming (sorry for the delay)...

**Comment by 2bndy5** [comment]
I can confirm this behavior with TMRh20's code running on 2 XIAO BLE boards as TX nodes while using my RPi4 as the RX node. Changing proximity of the node with failed transmissions helped slightly. Strangely enough, I'm also seeing `SPIException`s on my RPi4 (using SPIDEV driver), so I'll probably switch to a RP2040+nRF24L01 as the RX node (for a reliable base line).

I have other nRF5x boards (with SWD ports) that I can use to switch to a different Arduino core, namely the sandeep nRF5 core (for the microbit) and Adafruit's offshoot. I'll have to pick this up tomorrow.

**Comment by leleleoo123** [comment]
Yes, I also run into these compiler error and added `return true`

**Comment by leleleoo123** [comment]
I try to replaced the xiaoBle with adafruit feather nrf52840, same behavior.

**Comment by 2bndy5** [comment]
I forgot about this one, sorry. I'm still not sure where to begin; figured I'd try a different Arduino core to narrow the causes. If that still shows the problem, then I'd hit the datasheet again and compare code to the nRF5 SDK's ESB module...

**Comment by leleleoo123** [comment]
thank you so much!

**Comment by leleleoo123** [comment]
I have switched to nRF Connect SDK, based on the ptx & prx examples, I can successfully send packets from 4 xiao bles to one receiver.

**Comment by 2bndy5** [comment]
@TMRh20 Not sure if this issue should be closed as it is a known bug.

@leleleoo123 I've also had zephyr project support in RF24 on my todo list. Hopefully, that would make it easier to use nRF24L01 modules with the nRF Connect SDK.

**Comment by TMRh20** [comment]
Yeah, re-opening this as it is still a bug with this lib.

**Comment by 2bndy5** [comment]
If I had to guess, I think this the TX_READY event 
https://github.com/TMRh20/nrf_to_nrf/blob/95d994cd793481d97e3dfafc04070b75ade8554e/src/nrf_to_nrf.cpp#L334-L337
should be handled by the `write()` method instead of `stopListening()`. In my head this makes sense because the lib only uses 1 payload at a time (no need to keep CE high to empty a TX FIFO like we do for RF24).

I'm still digging in the datasheet, but I doubt that TX_IDLE is the same as nRF24's Standby-II mode. The behavior seen here seems to indicate that TX_IDLE is more like Standby-I mode.

**Comment by 2bndy5** [comment]
Um, so I tested this patch:
```diff
diff --git a/src/nrf_to_nrf.cpp b/src/nrf_to_nrf.cpp
index 4768daa..09d519b 100644
--- a/src/nrf_to_nrf.cpp
+++ b/src/nrf_to_nrf.cpp
@@ -231,6 +231,10 @@ bool nrf_to_nrf::write(void *buf, uint8_t len, bool multicast) {
     memcpy(&radioData[2], buf, len);

     // radioData[0] = ackPID++;
+    NRF_RADIO->EVENTS_TXREADY = 0;
+    NRF_RADIO->TASKS_TXEN = 1;
+    while (NRF_RADIO->EVENTS_TXREADY == 0);
+    NRF_RADIO->EVENTS_TXREADY = 0;
     NRF_RADIO->EVENTS_END = 0;
     NRF_RADIO->TASKS_START = 1;
     while (NRF_RADIO->EVENTS_END == 0) {
@@ -331,10 +335,6 @@ void nrf_to_nrf::stopListening(bool setWritingPipe, bool resetAddresses) {
   if (setWritingPipe) {
     NRF_RADIO->TXADDRESS = 0x00;
   }
-  NRF_RADIO->EVENTS_TXREADY = 0;
-  NRF_RADIO->TASKS_TXEN = 1;
-  while (NRF_RADIO->EVENTS_TXREADY == 0);
-  NRF_RADIO->EVENTS_TXREADY = 0;
   inRxMode = false;
 }
```

And now both my Xaio BLE boards (acting as nodes 1 & 3) are able to get through to my RPi4 (acting as the receiver). It _may still be_ a lucky guess. I'm not sure if the patch will have other side-effects on other examples.

Looking at the datasheet, I think the nRF52840 is designed to transmit a carrier wave while in TX_READY. My impression is that the carrier wave was drowning out other signals on the same channel.
![image](https://user-images.githubusercontent.com/14963867/223598734-07f971a9-2668-4aec-b219-a8ffbb2acd57.png)

- I've tried using my RF Explorer to show the ambient noise, but this was rather inconclusive without telling it to hold the detected peaks (which showed noise on channels 75-77 using a -64 dBM threshold).
- Running the RF24's scanner from examples_pico on my Feather RP2040 showed noise that seemed centered on channels 0x4B-0x4D (75-77). Remember that the scanner example works best when there's a preamble present, and (I highly suspect) the 52840's unsolicited carrier wave in TX_IDLE may not resemble/use a preamble.

**Comment by leleleoo123** [comment]
don't understand, but you guys rock

**Comment by TMRh20** [comment]
Did a quick test with RF24Ethernet & MQTT and it still works as normal with the patch. I'll run it through some more testing on my end as soon as I can.

**Comment by 2bndy5** [comment]
I'm not sure if [`startWrite()`](https://github.com/TMRh20/nrf_to_nrf/blob/95d994cd793481d97e3dfafc04070b75ade8554e/src/nrf_to_nrf.cpp#L274-L278) is actually usable as it is now. With the TX_READY event asserted in `write()` instead of `stopListening()`, I think TX_READY would also need to be asserted in `startWrite()`. However, that method's nature implies the use of interrupts or manually checking when the radio asserts the EVENTS_END flag... Otherwise, using `startWrite()` 
- with TX_READY asserted might instigate a similar issue
- as it is now but with TX_READY asserted only in `write()` might lead a misuse of the radio peripheral depending on whatever the last radio state was set. An excerpt from the datasheet:

  > The RADIO does not prevent a task from being triggered from the wrong state. If a task is triggered from the wrong state, for example if the RXEN task is triggered from the RXDISABLE state, this may lead to incorrect behaviour.

All this just to say: I think using shortcuts for the radio peripherals' READY_START (between READY and START events) and END_DISABLE (between END and DISABLE events) might be easier to avoid this issue or instigate another.

**Comment by TMRh20** [comment]
I've been giving this some thought, even though I haven't had much time to actually work on the issue, and I don't think there is much hope of getting `startWrite()` to work properly without interrupts. I just don't see how. I think it might be best to just make it call the normal write, so as not to break things completely, and put a note in documentation or something.

As far as the shortcuts go, that will take some time for me to understand how it will change things. I might just put your initial fix into play and then look at shortcuts later on.

**Comment by 2bndy5** [comment]
Yeah, interrupts will have to be Arduino core specific; ie. bare metal (like in [sandeepmistry/arduino-nRF5](https://github.com/sandeepmistry/arduino-nRF5)) vs some RTOS (like in [adafruit/nrf52](https://github.com/adafruit/Adafruit_nRF52_Arduino)). Bare metal _should_ be easier because you wouldn't need to contend with the complexity of a RTOS' tasks/semaphores.

**Comment by TMRh20** [comment]
So I finally had a chance to do some testing, and this patch seems to work OK from everything I can test so I committed the changes. I haven't played around with `startWrite()` but we can consider the function broken for now I guess.

**Comment by 2bndy5** [comment]
> I haven't played around with startWrite() but we can consider the function broken for now I guess.

Could be a new issue (for tracking purposes)

I suppose this issue is adequately solved then. Did you want me to run some more tests with both of my XIAO BLE boards?

**Comment by 2bndy5** [comment]
I think this issue was rather specific to the Multiceiver feature/example because it is the only example/usage that I can think of where we have the radio idle in TX.

**Comment by TMRh20** [comment]
Yeah, I'll open a new issue for `startWrite()`, I think we're pretty much good. We can always re-open if any issues emerge.
